README.md
loadmyenvironment (v1.1.0)
[![Build Status](https://travis-ci.org/deepaknverma/loadmyenvironment.svg?branch=master)](https://travis-ci.org/deepaknverma/loadmyenvironment)  
============
A NodeJS module to fetch application related variables from and configuration object either from a file, url or a dependency module. 

OVERVIEW
=========
The best was to have global environment variable is to set them in OS environment. But it is not safe enough. if someone has access to your machine, it can open up a pandora box to access other services in your application network. I wanted a safe way to provide these information to my application. Load environment is capable of attaining below goals:

- Able to load environment from a file path
- Able to load environment from a URL
- Able to either return the variables or set them in process environment.
- Able to decrypt an encrypted configuration file and recursively find information your app need

REQUIRE
========
- returnConfig: if set `true` then return result else set them in process environment
- functionKey: array of all the keys you want to be returned from the configuration. if only one key required then pass as a string
- configPath: configuration path. Can be file path or an url. **Note** *configuration should be an object*
- encrypted: if the file is encrypted?
- key: password to decrypt the configuration

USAGE:
========

see testcases on [github](https://github.com/deepaknverma/loadmyenvironment)

TODO
=====
- currently only support `aes-256-cbc` encryption. need to add more algorithm support




CHANGELOG
[18-10-2015]
=========
 - added support to read from url and process recursively
 - ability to return value or set in the process env
 - ability to pass the key for which you want to return value

[19-10-2015]
=========
- Added test cases.
- Ability to pass custom path to your own env file

test/index.js
/*
* @Author: Deepak Verma
* @Date:   2015-11-19 12:25:27
* @Last Modified by:   Deepak Verma
* @Last Modified time: 2015-11-19 12:25:46
*/

'use strict';

require('./unit');


test/unit/index.js
/*
* @Author: Deepak Verma
* @Date:   2015-11-19 12:25:14
* @Last Modified by:   dverma
* @Last Modified time: 2017-06-06 12:12:54
*/

'use strict';
const should = require('should');
const config = require('../../');
const nock = require('nock');
const path = require('path');
const fs = require('fs');
const config_path = path.join(__dirname, '/../resources/config.env');

describe('Enviornment Variable', () => {

	beforeEach((done) => {
		nock('http://localhost')
			.get('/env.json')
			.replyWithFile(200, __dirname + '/../resources/env.json');

		nock('http://localhost')
			.defaultReplyHeaders({
				'Content-Type': 'text/plain'
			})
			.get('/config.env')
			.replyWithFile(200, __dirname + '/../resources/config.env');
		nock('http://localhost')
			.get('/env.js')
			.reply(404);
		done();
	});

	it('should return api object with a file path', (done) => {
		config({
			returnConfig: true,
			functionKey: 'apis',
			configPath: config_path,
			encrypted: true,
			key: 'thi5i5apa55word'
		})
			.then(data => {
				// console.log(data);
				should.equal(typeof data, 'object', 'It is returning an Object');
				should.equal(data.apis.postRequest.emails.noreply, 'noreply@localhost', 'It have value noreply@localhost');
				done();
			})
			.catch(err => {
				console.log(err);
				should.not.exist(err);
				done();
			});
	});

	it('should return true with a file path', (done) => {
		// setting up the environment variable
		process.env.NODE_ENV = 'dev';		
		config({
			returnConfig: false,
			functionKey: 'apis',
			configPath: config_path,
			encrypted: true,
			key: 'thi5i5apa55word'
		})
			.then(data => {
				// console.log(data);
				should.equal(data.apis, true, 'It is returning Bool');
				done();
			})
			.catch(err => {
				console.log(err);
				should.not.exist(err);
				done();
			});
	});

	it('should load the env file in process.env', (done) => {
		process.env.NODE_ENV = 'dev';
		process.env.APP_CONFIG_PATH = path.join(__dirname, '/../resources/config.env');
		config({
			returnConfig: false,
			encrypted: true,
			key: 'thi5i5apa55word'
		})
			.then(data => {
				// console.log(data);
				should.equal(data, true, true);
				should.exist(process.env.NODE_ENV);
				should.exist(process.env.APP_CONFIG_PATH);
				done();
			})
			.catch(err => {
				// console.log(err);
				should.not.exist(err);
				done();
			});
	});

	it('should return true with a file path set in server env', (done) => {
		process.env.NODE_ENV = 'dev';
		process.env.APP_CONFIG_PATH = path.join(__dirname, '/../resources/config.env');
		config({
			returnConfig: false,
			functionKey: 'apis',
			encrypted: true,
			key: 'thi5i5apa55word'
		})
			.then(data => {
				// console.log(data);
				should.equal(data.apis, true, 'It is returning Bool');
				done();
			})
			.catch(err => {
				// console.log(err);
				should.not.exist(err);
				done();
			});
	});

	it('should return apis object with a URL', (done) => {
		config({
			returnConfig: true,
			functionKey: 'apis',
			configPath: 'http://localhost/env.json'
		})
			.then(data => {
				should.equal(typeof data, 'object', 'It is returning an Object');
				done();
			})
			.catch(err => {
				// console.log('err: ', err);
				should.not.exist(err);
				done();
			});
	});

	it('should return apinotifications object with a URL when function keys is an array', (done) => {
		config({
			returnConfig: true,
			functionKey: ['email', 'apiNotifications'],
			configPath: 'http://localhost/env.json'
		})
			.then(data => {
				// console.log(data);
				should.equal(typeof data, 'object', 'It is returning an Object');
				done();
			})
			.catch(err => {
				// console.log('err: ', err);
				should.not.exist(err);
				done();
			});
	});

	it('should return apinotifications object with a URL when functionkeys is a single string', (done) => {
		config({
			returnConfig: true,
			functionKey: 'apiNotifications',
			configPath: 'http://localhost/env.json'
		})
			.then(data => {
				// console.log(data);
				should.equal(typeof data, 'object', 'It is returning an Object');
				done();
			})
			.catch(err => {
				// console.log('err: ', err);
				should.not.exist(err);
				done();
			});
	});

	it.skip('should return api object with a URL when encrypted', (done) => {
		config({
			returnConfig: false,
			functionKey: 'apis',
			configPath: 'http://localhost/config.env',
			encrypted: true,
			key: 'thi5i5apa55word'
		})
			.then(function(data) {
				// console.log(data);
				should.equal(data.apis, true, 'It is returning Bool');
				done();
			})
			.catch(function(err) {
				// console.log(err);
				should.not.exist(err);
				done();
			});
	});

	// NOTE: FIXME: need to revist this case
	it('should return error when key does not exist', function(done) {
		config({
			returnConfig: false,
			functionKey: 'badkey',
			configPath: 'http://localhost/env.json'
		})
			.then(function(data) {
				// console.log(data);
				should.not.exst(data);
				should.equal(data, true, 'It is returning Bool');
				done();
			})
			.catch(function(err) {
				// console.log(err);
				should.exist(err);
				done();
			});
	});

	it('should return object with a URL when array of function keys are passed', function(done) {
		config({
			returnConfig: true,
			functionKey: ['app-port', 'schema', 'proxy'],
			configPath: 'http://localhost/env.json'
		})
			.then(function(data) {
				// console.log(data);
				should.exist(data);
				should.equal(typeof data, 'object', 'It is returning object');
				should.equal(data['app-port'], 3000);
				should.equal(typeof data.proxy, 'object');
				done();
			})
			.catch(function(err) {
				// console.log(err);
				should.not.exist(err);
				done();
			});
	});

	it('should return true with a URL when array of function keys are passed', function(done) {
		config({
			returnConfig: false,
			functionKey: ['app-port', 'proxy'],
			configPath: 'http://localhost/env.json'
		})
			.then(function(data) {
				// console.log(data);
				should.equal(data.proxy, true, 'It is returning Bool');
				should.equal(data['app-port'], true, 'It is returning Bool');
				done();
			})
			.catch(function(err) {
				should.not.exist(err);
				// console.log(err);
				done();
			});
	});

	it('should return Object with a URL when functionKey is not passed', function(done) {
		config({
			returnConfig: true,
			configPath: 'http://localhost/env.json'
		})
			.then(function(data) {
				// console.log(data);
				should.equal(typeof data, 'object', 'It is returning an Object');
				should.equal(data['app-port'], 3000, 'It has value for the port');
				done();
			})
			.catch(function(err) {
				should.not.exist(err);
				// console.log(err);
				done();
			});
	});

	it('should return error a for a URL bad json file', function(done) {
		config({
			returnConfig: true,
			configPath: 'http://localhost/env.js'
		})
			.then(function(data) {
				// console.log(data);
				should.not.exist(data);
				done();
			})
			.catch(function(err) {
				should.exist(err);
				// console.log(err);
				done();
			});
	});
//
	it.skip('should return object with a file path when functionKey is not passed', function(done) {
		config({
			returnConfig: true,
			configPath: config_path
		})
			.then(function(data) {
				// console.log(data);
				should.equal(typeof data, 'object', 'It is returning an Object');
				done();

			})
			.catch(function(err) {
				should.not.exist(err);
				// console.log(err);
				done();
			});
	});

	it('should throw an error', function(done) {
		config()
			.then(function(data) {
				should.not.exist(data);
				done();
			})
			.catch(function(err) {
				should.equal(err.title, 'ConfigError', err.title);
				done();
			});
	});

	it('should throw an error invalid config path', function(done) {
		config({
			returnConfig: true,
			configPath: '/bad/path/env.json'
		})
			.then(function(data) {
				// console.log(data);
				should.not.exist(data);
				done();
			})
			.catch(function(err) {
				// console.log(err)
				should.exist(err);
				should(err).have.property('code', 'MODULE_NOT_FOUND');
				done();
			});
	});

	it('should throw an error bad json', function(done) {
		config({
			returnConfig: true,
			configPath: path.join(__dirname, '../data/bad-sample.json')
		})
			.then(function(data) {
				// console.log(data);
				should.not.exist(data);
				done();
			})
			.catch(function(err) {
				// console.log(err)
				should.exist(err);
				// should(err).have.property('code','MODULE_NOT_FOUND');
				done();
			});
	});


	it('should return error for invalid key', function(done) {
		config({
			returnConfig: true,
			functionKey: 'database',
			configPath: path.join(__dirname, '/../resources/config.env'),
			encrypted: true,
			key: 'xxx'
		})
			.then(function(data) {
				// console.log(data);
				should.not.exist(data);
				done();

			})
			.catch(function(err) {
				// console.log(err);
				should.exist(err);
				should(err.title).equal('ConfigError');
				done();
			});
	});
	afterEach(() => {
		nock.cleanAll();
	});
});




index.js
/*
* @Author: Deepak Verma
* @Date:   2015-10-26 18:12:08
* @Last Modified by:   dverma
* @Last Modified time: 2017-06-06 12:11:49
*/

'use strict';
const isURL = require('is-url');
const request = require('request-promise');
const errs = require('errs');
const htauCrypt = require('encryptit').crypt;
const _ = require('underscore');
const tmp = require('tmp');
const fs = require('fs');
const node_env = process.env.NODE_ENV || 'dev';
let payload, returnValue = {};
/**
 * [recursiveIteration recurively search for needle in the object]
 * @param  {[object]} object      [object to search from]
 * @param  {[string]} functionKey [needle to search]
 * @return {[string]}             [return object value if found else bool false]
 */
const recursiveIteration = (object, functionKey) => {
	for (var property in object) {
		if (object.hasOwnProperty(property)) {
			if (property === functionKey) {
				payload = object[property];
			}
			else if (typeof object[property] === 'object') {
				recursiveIteration(object[property], functionKey);
			}
		}
	}
};

const processEnv = (env, options) => {
	return new Promise((success, fail) => {
		if (typeof env === 'object') {
			if (options.functionKey) {
				const functionkeys = options.functionKey;
				if (Array.isArray(functionkeys) && functionkeys.length) {
					for (let i = 0, len = functionkeys.length; i < len; i++) {
						recursiveIteration(env, functionkeys[i]);
						if (options.returnConfig === true) {
							returnValue[functionkeys[i]] = payload;
						} else {
							process.env[functionkeys[i]] = payload;
							returnValue[functionkeys[i]] = true;
						}
					}
					return success(returnValue);
				} else {
					recursiveIteration(env, functionkeys);
					if (options.returnConfig === true) {
						returnValue[functionkeys] = payload;
					} else {
						process.env[functionkeys] = payload;
						returnValue[functionkeys] = true;
					}
				}
				return success(returnValue);
			} else {
				if (options.returnConfig === true) {
					return success(env);
				} else {
					process.env[node_env] = env;
					return success(true);
				}
			}
		} else {
			return fail(errs.create({
				title: 'ConfigError',
				message: 'unable to process environment file',
				options: options
			}));
		}
	});
};

/**
 * [config pulls Config out of the path set in env]
 * @param  {[object]} options [{return: bool, functionKey: key name to be retrieved}]
 * @return {[object]}         [description]
 * @sample Path 	//var path = 'https://gist.githubusercontent.com/deepaknverma/6119f4e95730e6533fd6/raw/43d36d1058e24e2cbbfeb3f5b763d4ddcbc613fb/sample.json';
 *         			//var path = './env.json';
 */
const config = (options) => {
	return new Promise((success, fail) => {
		let config, errors, jsonConfig, env, requestOptions;
		// We need configuration to work with
		if (!options) {
			errors = errs.create({
				title: 'ConfigError',
				message: 'No options supplied',
				options: null
			});
			return fail(errors);
		}

		// default returnConfig to true
		options.returnConfig = (options.returnConfig !== undefined && typeof options.returnConfig === 'boolean') ? options.returnConfig : true;

		// Lets see if the user has passed some config file else we will user the env variable
		if (options.configPath) {
			config = options.configPath;
		} else if (process.env.APP_CONFIG_PATH) {
			config = process.env.APP_CONFIG_PATH;
		}

		// check if path or url
		if (isURL(config)) {
			requestOptions = {
				uri: config,
				json: true,
				resolveWithFullResponse: true
			};
			request(requestOptions)
				.then((response) => {
					const responseBody = response.body;
					if (options.encrypted && options.key) {
						const tmpobj = tmp.fileSync({ postfix: '.env' });
						fs.writeFileSync(tmpobj.name, responseBody, { encoding: 'UTF-8' });
						const crypt = new htauCrypt({
							bytes: 256,
							algorithmSuffix: 'cbc',
							key: options.key,
							inputFilePath: tmpobj.name,
							return: true
						});
						return crypt.decryptCBC();

					} else {
						return responseBody;
					}
				})
				.then((data) => {
					tmp.setGracefulCleanup();
					jsonConfig = (options.encrypted && options.key) ? JSON.parse(data) : data;
					env = _.extend(jsonConfig[node_env], jsonConfig.base);
					return processEnv(env, options);
				})
				.then((data) => {
					return success(data);
				})
				.catch((err) => {
					// console.error('URL Error: ', err);
					errors = errs.create({
						title: 'ConfigError',
						message: err,
						options: options
					});
					return fail(errors);
				});
		} else {
			try {
				if (options.encrypted && options.key) {
					const crypt = new htauCrypt({
						bytes: 256,
						algorithmSuffix: 'cbc',
						key: options.key,
						inputFilePath: config,
						return: true
					});
					crypt.decryptCBC()
						.then((data) => {
							jsonConfig = JSON.parse(data);
							env = _.extend(jsonConfig[node_env], jsonConfig.base);
							return processEnv(env, options);
						})
						.then((data) => {
							return success(data);
						})
						.catch((err) => {
							const error = new Error(err);
							return fail(errs.create({
								title: 'ConfigError',
								code: 10109,
								message: (error.message) ? error.message : 'Env Configuration Error',
								options: null
							}));
						});
				} else {
					jsonConfig = require(config);
					env = _.extend(jsonConfig[node_env], jsonConfig.base);
					processEnv(env, options)
						.then((data) => {
							return success(data);
						})
						.catch((err) => {
							return fail(err);
						});
				}
			}
			catch (err) {
				return fail({ code: 'MODULE_NOT_FOUND', message: 'Unable to require config file', trace: err });
			}

		}

	});
};

module.exports = config;
